### ISSUE TITLE
Could not decode the IBM420_rtl encoded file

### ISSUE THREAD

**Issue by adoyle-h** [editorconfig-checker/editorconfig-checker#252]
Type `printf 'inc\nsub\n' > abc` to create a file `abc`.
There is no `.ecrc` and type `editorconfig-checker`. It shows error. Why?

<img width="503" alt="CleanShot 2023-02-07 at 15 22 27@2x" src="https://user-images.githubusercontent.com/1998490/217175826-4ae07360-28e9-4dd7-8ea6-c27d8933b9e8.png">

**Comment by oschwald** [comment]
I am also seeing this on certain ASCII files.

**Comment by polarathene** [comment]
`sub` as the file content is sufficient for your reproduction it seems.

---

## Reproduction

Reproduction via the Docker image:

```console
$ docker run --rm --tty \
    --volume "${PWD}:/ci:ro" \
    --workdir "/ci" \
    'mstruebing/editorconfig-checker:2.7.1' \
    ec -config "/ci/test/linting/.ecrc.json"

Config: {ShowVersion:false Help:false DryRun:false Path:/ci/test/linting/.ecrc.json Version: Verbose:false Debug:true IgnoreDefaults:true SpacesAftertabs:false NoColor:false Exclude:[^test/bats/ ^test/test_helper/bats-(assert|support) \.git/] AllowedContentTypes:[text/ application/octet-stream application/ecmascript application/json application/x-ndjson application/xml +json +xml] PassedFiles:[] Disable:{EndOfLine:false Indentation:false InsertFinalNewline:false TrimTrailingWhitespace:false IndentSize:false MaxLineLength:false} Logger:{Verbosee:false Debugg:true NoColor:false}}
AddToFiles: filePath: /ci/test.txt, contentType: text/plain
Could not decode the IBM420_rtl encoded file: /ci/test.txt
unrecognized charset IBM420_rtl
```

**`.ecrc.json`** (_config used, **not required** to reproduce_):
```json
{
  "Debug": true,
  "IgnoreDefaults": true,
  "Exclude": [
    "^test/bats/",
    "^test/test_helper/bats-(assert|support)",
    "\\.git/"
  ]
}
```

File that will fail:

```
sub
```

`c2Vj` and `caV` both trigger it too, as does much longer file content like below.

---

## Investigation

### Upstream (`chardet`) is responsible for detecting this charset

Looks like a package for the detection is used for this charset, and it does so [via a lookup table with the char (byte)](https://github.com/baulk/chardet/blame/462a6931549f8c787f3251fdec92a6b7616544ea/single_byte.go#L833-L851) as the key. It was [added to `chardet` in 2012](https://github.com/baulk/chardet/commit/5c9192840eaa32b2269c6adacb48a8689191a61b) with little details to reference beyond that (_not much changes to the file since_).

That table covers the equivalent regex range that you can plug into `grep` (_`-P` for perl regex, which supports the hex notation_):

```bash
grep -P '[\x40,\x42-\x49,\x51,\x52,\x55-\x59,\x62-\x69,\x70-\x79,\
x80-\xA0,\xA2-\xB5,\xB8-\xBF,\xCB,\xCD,\xCF,\xDA-\xDF,\xEA-\xEF,\xFB-\xFE]' filename_here
```

In my file that triggered this (_the gibberish is base64 encoded, this is for `SMTP` tests_):

```
EHLO mail
AUTH LOGIN
c29tZS51c2VyQGxvY2FsaG9zdC5sb2NhbGRvbWFpbg==
c2VjcmV0
QUIT
```

we can see over 50% is matched (highlighted red):

![image](https://github.com/editorconfig-checker/editorconfig-checker/assets/5098581/fb5714f5-a674-44e9-a271-3e3b1b28d812)

Many other characters when used as index to the lookup table would map to `0x40`. There is an [array of 64 elements for `IBM420_rtl`](https://github.com/baulk/chardet/blame/462a6931549f8c787f3251fdec92a6b7616544ea/single_byte.go#L853-L858)
that provides a sequence of 3 bytes matching those in the lookup table values. Not quite sure exactly why `caV` is sufficient to match this encoding, while `c2V` is not, but `c2Vj` is. These chars that change should all map to `0x40`? So something else is involved there. Might be based on hit rate of content with ngrams 🤷‍♂️ 

### `IBM420_rtl` as an encoding result in `editorconfig-checker` seems unintentional

Pretty sure that's the culprit regardless. It's usage can be found in `editorconfig-checker` from [this commit](https://github.com/editorconfig-checker/editorconfig-checker/commit/9d4703a385c3467d6c63e3c74cb11836a9d543a5#diff-3339fd5eeca44afac33a7719bbf4456580c75fed67e81d6d2eb74bb62ef01e1aR117-R137) that introduced it. That commit message mentions:

> _`chardet`'s `DetectBest()` returns multiple charsets with the same Confidence level. Unfortunately, it sometimes returns a different charset._
> _This commit fixes this by sorting the results returned by `DetectAll()` and returning the first alphabetically._
> _This could certainly be improved._

Since the encoding is not supported, we hit [this error](https://github.com/editorconfig-checker/editorconfig-checker/commit/9d4703a385c3467d6c63e3c74cb11836a9d543a5#diff-3339fd5eeca44afac33a7719bbf4456580c75fed67e81d6d2eb74bb62ef01e1aR142) as it's not [one of the accepted encodings](https://github.com/editorconfig-checker/editorconfig-checker/commit/9d4703a385c3467d6c63e3c74cb11836a9d543a5#diff-3339fd5eeca44afac33a7719bbf4456580c75fed67e81d6d2eb74bb62ef01e1aR23), but it's still being [returned from here](https://github.com/editorconfig-checker/editorconfig-checker/commit/9d4703a385c3467d6c63e3c74cb11836a9d543a5#diff-3339fd5eeca44afac33a7719bbf4456580c75fed67e81d6d2eb74bb62ef01e1aR133) where it gets sorted after. It probably should have filtered to the accepted encodings only, since if there is more than 1 in these results that's still acceptable? `utf8` is probably there, just sorting approach would prioritize other encodings found. Sorting by the confidence might have been more reliable (_assuming `utf8` is there and had a higher confidence level_).

This was contributed in June 2022 for https://github.com/editorconfig-checker/editorconfig-checker/pull/206 for `v2.6.0`, and [has had no changes since](https://github.com/editorconfig-checker/editorconfig-checker/blob/9d4703a385c3467d6c63e3c74cb11836a9d543a5/pkg/encoding/encoding.go).

---

## Fixing it

I am not a Go dev, so that's about as much detective work as I can do 😅 

@rasa or @mstruebing were both involved in the June 2022 PR for this change, they may know how take those findings above and fix it.

Probably just requires filtering out unsupported encodings before sorting, and failing if no supported encodings remain?

**Comment by rasa** [comment]
@adoyle-h @oschwald @polarathene 
Thank you for the detailed error report. It made it extremely easy to craft the fix. PR coming soon!

**Comment by oschwald** [comment]
Thank you for fixing this!
